File: backend/services/history_compression_service.py
"""
History compression service.

Compresses history.md entries into consolidated_history.md using the
History_Summarizer agent. Groups turns into batches and generates
meaningful subtitles for each consolidated section.
"""

import logging
import re
from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional

from infrastructure.database import models
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select

from services.world_service import WorldService

if TYPE_CHECKING:
    from sdk import AgentManager

logger = logging.getLogger("HistoryCompressionService")

# Pattern to match turn entries in history.md
TURN_PATTERN = re.compile(r"^## Turn (\d+) - (.+)$", re.MULTILINE)

# Default batch size (number of turns to compress into one section)
BATCH_SIZE = 3


@dataclass
class TurnEntry:
    """A single turn entry from history.md."""

    turn_number: int
    location: str
    content: str


@dataclass
class CompressionResult:
    """Result of history compression."""

    success: bool
    turns_compressed: int
    sections_created: int
    message: str


def _parse_history_into_turns(history_content: str) -> list[TurnEntry]:
    """
    Parse history.md content into individual turn entries.

    Args:
        history_content: Raw content of history.md

    Returns:
        List of TurnEntry objects
    """
    entries = []

    # Find all turn headers and their positions
    matches = list(TURN_PATTERN.finditer(history_content))

    for i, match in enumerate(matches):
        turn_number = int(match.group(1))
        location = match.group(2)

        # Get content between this header and the next (or end of file)
        start = match.end()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(history_content)

        content = history_content[start:end].strip()

        entries.append(TurnEntry(turn_number=turn_number, location=location, content=content))

    return entries


def _group_turns_into_batches(turns: list[TurnEntry], batch_size: int = BATCH_SIZE) -> list[list[TurnEntry]]:
    """
    Group turn entries into batches for compression.

    Args:
        turns: List of turn entries
        batch_size: Number of turns per batch

    Returns:
        List of batches (each batch is a list of TurnEntry)
    """
    batches = []
    for i in range(0, len(turns), batch_size):
        batch = turns[i : i + batch_size]
        batches.append(batch)
    return batches


def _format_batch_for_summarizer(batch: list[TurnEntry]) -> str:
    """
    Format a batch of turns for the History_Summarizer agent.

    Args:
        batch: List of turn entries to summarize

    Returns:
        Formatted text for the agent
    """
    lines = []
    for entry in batch:
        lines.append(f"## Turn {entry.turn_number} - {entry.location}")
        lines.append(entry.content)
        lines.append("")

    return "\n".join(lines)


async def _get_history_summarizer_agent(db: AsyncSession) -> Optional[models.Agent]:
    """
    Get the History_Summarizer agent from the database.

    Returns:
        Agent model or None if not found
    """
    result = await db.execute(select(models.Agent).where(models.Agent.group == "gameplay"))
    agents = result.scalars().all()

    for agent in agents:
        if agent.name == "History_Summarizer":
            return agent

    logger.warning("History_Summarizer agent not found in database")
    return None


async def _generate_compressed_section(
    db: AsyncSession,
    agent_manager: "AgentManager",
    batch: list[TurnEntry],
    room_id: int = 0,
) -> Optional[str]:
    """
    Generate a compressed section for a batch of turns using History_Summarizer.

    Args:
        db: Database session
        agent_manager: AgentManager instance
        batch: List of turn entries to compress
        room_id: Room ID for task tracking (default 0)

    Returns:
        Compressed section text (including ## [subtitle] header), or None on failure
    """
    # Import locally to avoid circular imports
    from domain.entities.agent import AgentConfigData
    from domain.value_objects import TaskIdentifier
    from domain.value_objects.contexts import AgentResponseContext

    from services import AgentConfigService
    from services.prompt_builder import build_system_prompt

    summarizer = await _get_history_summarizer_agent(db)
    if not summarizer:
        return None

    try:
        # Load agent config from filesystem
        config_data = AgentConfigData()
        if summarizer.config_file:
            loaded_config = AgentConfigService.load_agent_config(summarizer.config_file)
            if loaded_config:
                config_data = loaded_config

        # Build system prompt
        system_prompt = build_system_prompt(summarizer.name, config_data)

        # Format the batch content
        batch_content = _format_batch_for_summarizer(batch)

        # Build user message
        user_message = f"""Please compress the following turn entries into a single consolidated section.

## Turn Entries to Compress
{batch_content}

## Instructions
1. Create a meaningful subtitle in square brackets that captures the key event/theme
2. Write a concise summary that preserves important events, characters, and outcomes
3. Output ONLY the consolidated section in this format:

## [meaningful_subtitle_here]
Your consolidated summary here..."""

        # Build AgentResponseContext
        task_id = TaskIdentifier(room_id=room_id, agent_id=summarizer.id)
        context = AgentResponseContext(
            system_prompt=system_prompt,
            user_message=user_message,
            agent_name=summarizer.name,
            config=config_data,
            room_id=room_id,
            agent_id=summarizer.id,
            group_name=summarizer.group,
            task_id=task_id,
        )

        # Generate summary via AgentManager
        response_text = ""
        async for event in agent_manager.generate_sdk_response(context):
            if event.get("type") == "content_delta":
                response_text += event.get("delta", "")
            elif event.get("type") == "stream_end":
                if event.get("response_text"):
                    response_text = event["response_text"]

        logger.info(f"History_Summarizer generated section: {response_text[:100]}...")
        return response_text.strip() if response_text else None

    except Exception as e:
        logger.error(f"Error generating compressed section: {e}")
        import traceback

        traceback.print_exc()
        return None


class HistoryCompressionService:
    """Service for compressing world history."""

    @staticmethod
    async def compress_history(
        db: AsyncSession,
        world_name: str,
        agent_manager: "AgentManager",
        batch_size: int = BATCH_SIZE,
    ) -> dict:
        """
        Compress history.md into consolidated_history.md.

        Args:
            db: Database session
            world_name: Name of the world
            agent_manager: AgentManager instance
            batch_size: Number of turns per compressed section

        Returns:
            Dict with success, turns_compressed, sections_created, message
        """
        # Load history.md
        history_content = WorldService.load_history(world_name)

        if not history_content or history_content.strip() == "# World History":
            return {
                "success": True,
                "turns_compressed": 0,
                "sections_created": 0,
                "message": "No history to compress",
            }

        # Parse into turns
        turns = _parse_history_into_turns(history_content)

        if not turns:
            return {
                "success": True,
                "turns_compressed": 0,
                "sections_created": 0,
                "message": "No turn entries found in history",
            }

        # Group into batches
        batches = _group_turns_into_batches(turns, batch_size)

        logger.info(f"Compressing {len(turns)} turns into {len(batches)} batches for world '{world_name}'")

        # Generate compressed sections for each batch
        compressed_sections = []
        for batch in batches:
            section = await _generate_compressed_section(db, agent_manager, batch)
            if section:
                compressed_sections.append(section)
            else:
                logger.warning(f"Failed to compress batch starting at turn {batch[0].turn_number}")

        if not compressed_sections:
            return {
                "success": False,
                "turns_compressed": 0,
                "sections_created": 0,
                "message": "Failed to generate any compressed sections",
            }

        # Append to consolidated_history.md
        world_path = WorldService.get_world_path(world_name)
        consolidated_file = world_path / "consolidated_history.md"

        # Read existing content if file exists
        existing_content = ""
        if consolidated_file.exists():
            with open(consolidated_file, "r", encoding="utf-8") as f:
                existing_content = f.read()

        # Append new sections
        new_content = "\n\n".join(compressed_sections)
        if existing_content:
            final_content = existing_content.rstrip() + "\n\n" + new_content
        else:
            final_content = new_content

        with open(consolidated_file, "w", encoding="utf-8") as f:
            f.write(final_content)

        logger.info(f"Written {len(compressed_sections)} sections to consolidated_history.md")

        # Clear history.md (keep header only)
        history_file = world_path / "history.md"
        with open(history_file, "w", encoding="utf-8") as f:
            f.write("# World History\n\n")

        # Invalidate history cache
        from services.world_service import _history_cache

        if world_name in _history_cache:
            del _history_cache[world_name]

        logger.info(f"Cleared history.md for world '{world_name}'")

        return {
            "success": True,
            "turns_compressed": len(turns),
            "sections_created": len(compressed_sections),
            "message": f"Compressed {len(turns)} turns into {len(compressed_sections)} sections",
        }

    @staticmethod
    def load_consolidated_history(world_name: str) -> str:
        """
        Load consolidated_history.md content.

        Args:
            world_name: Name of the world

        Returns:
            Content of consolidated_history.md or empty string
        """
        world_path = WorldService.get_world_path(world_name)
        consolidated_file = world_path / "consolidated_history.md"

        if not consolidated_file.exists():
            return ""

        with open(consolidated_file, "r", encoding="utf-8") as f:
            return f.read()

    @staticmethod
    def get_history_subtitles(world_name: str) -> list[str]:
        """
        Get list of subtitles from consolidated_history.md.

        Uses the same pattern as memory_parser for consistency.

        Args:
            world_name: Name of the world

        Returns:
            List of subtitle strings (without brackets)
        """
        content = HistoryCompressionService.load_consolidated_history(world_name)
        if not content:
            return []

        # Pattern matches ## [subtitle] headers
        subtitle_pattern = re.compile(r"^##\s*\[([^\]]+)\]", re.MULTILINE)
        matches = subtitle_pattern.findall(content)
        return matches

    @staticmethod
    def get_history_by_subtitle(world_name: str, subtitle: str) -> Optional[str]:
        """
        Get history content by subtitle.

        Args:
            world_name: Name of the world
            subtitle: The subtitle to retrieve (without brackets)

        Returns:
            Content of the history section, or None if not found
        """
        content = HistoryCompressionService.load_consolidated_history(world_name)
        if not content:
            return None

        # Parse into sections (same approach as memory_parser)
        subtitle_pattern = re.compile(r"^##\s*\[([^\]]+)\]", re.MULTILINE)
        matches = list(subtitle_pattern.finditer(content))

        for i, match in enumerate(matches):
            if match.group(1) == subtitle:
                # Get content between this header and the next (or end of file)
                start = match.end()
                end = matches[i + 1].start() if i + 1 < len(matches) else len(content)
                return content[start:end].strip()

        return None

File: backend/src/services/history-compression-service.ts
/**
 * Compressing `history.md` into `consolidated_history.md`.
 *
 * Ported from `backend/services/history_compression_service.py`.
 *
 * A world's `history.md` grows by one `## Turn N - Location` entry per travel,
 * and the Action Manager's context carries it whole. Left alone it eventually
 * dominates the prompt. Compression folds it into batches of three turns, hands
 * each batch to the `History_Summarizer` agent, and appends the summaries to
 * `consolidated_history.md` under `## [subtitle]` headings — the same shape as
 * an agent's `consolidated_memory.md`, which is what lets the `recall_history`
 * tool offer the subtitles and fetch one section on demand.
 *
 * ## The summarizer is injected
 *
 * The parsing, batching and formatting is the part that can be wrong in a way
 * that loses a player's history, and none of it needs a model. It is therefore
 * separated from the one call that does: {@link Summarize} is the whole of this
 * module's dependency on the SDK, exactly as `TapeExecutor` depends on a
 * `respond` function rather than on the turn runner. {@link createSummarizer}
 * supplies the real one.
 *
 * ## What lives elsewhere
 *
 * Python's three read helpers — `load_consolidated_history`,
 * `get_history_subtitles`, `get_history_by_subtitle` — are pure reads of a file
 * in the world directory and are already ported onto `WorldService`, next to
 * `loadHistory`. They are not duplicated here; call them there.
 */

import { existsSync, readFileSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'

import { and, eq } from 'drizzle-orm'

import { getSettings } from '../config/settings'
import type { Db } from '../db'
import { agents, type Agent } from '../db/schema'
import { getLogger } from '../infrastructure/logging/logger'
import { TurnRunner } from '../sdk/agent/turn-runner'
import type { SessionPool } from '../sdk/client/session-pool'
import { AgentConfigService } from './agent-config-service'
import { buildSystemPrompt } from './prompt-builder'
import { MtimeCache, WorldService } from './world-service'

const logger = getLogger('HistoryCompressionService')

// ============================================================================
// Constants
// ============================================================================

/**
 * `## Turn 12 - tavern` headings in `history.md`.
 *
 * `[\s\S]` is never involved: the pattern is line-anchored with `m` and no `s`,
 * so the location half stops at the newline exactly as Python's `.+` under
 * `MULTILINE` does.
 */
const TURN_PATTERN = /^## Turn (\d+) - (.+)$/gm

/** Turns folded into one consolidated section. */
export const BATCH_SIZE = 3

/** What `WorldService.createWorld` writes into a fresh `history.md`. */
const EMPTY_HISTORY_MARKER = '# World History'

/** The name the summarizer agent is registered under. */
const SUMMARIZER_NAME = 'History_Summarizer'

/** Group the summarizer belongs to; the lookup is scoped to it, as in Python. */
const SUMMARIZER_GROUP = 'gameplay'

// ============================================================================
// Types
// ============================================================================

/** One `## Turn N - Location` entry and the prose beneath it. */
export interface TurnEntry {
  turnNumber: number
  location: string
  /** Body between this heading and the next, trimmed. */
  content: string
}

/**
 * The compression route's response body.
 *
 * Keys are snake_case because this object *is* the JSON the endpoint returns
 * (`routers/game/worlds.py:484-486` returns the dict unchanged), and the parity
 * contract freezes the wire shape. It is the one place in this file where the
 * on-disk/on-wire spelling wins over the TypeScript one.
 */
export interface CompressionResult {
  success: boolean
  turns_compressed: number
  sections_created: number
  message: string
}

/**
 * The SDK call, narrowed to what compression needs.
 *
 * Returns the consolidated section — heading included — or `null` when the model
 * produced nothing. A rejection is a failure of that one batch, not of the run:
 * see {@link HistoryCompressionService.compressHistory}.
 */
export type Summarize = (request: SummarizeRequest) => Promise<string | null>

export interface SummarizeRequest {
  agent: Agent
  systemPrompt: string
  userMessage: string
  /** Room the work is attributed to; 0 outside a room, as in Python. */
  roomId: number
}

// ============================================================================
// Parsing / batching — no model involved
// ============================================================================

/**
 * Split `history.md` into its turn entries.
 *
 * Anything before the first heading — the `# World History` title, and any
 * hand-written preamble — is dropped, which is what makes the file's title
 * survive the rewrite at the end of a compression run rather than being folded
 * into the first summary.
 *
 * A malformed or headingless file yields `[]` rather than one giant entry, and
 * `compressHistory` treats that as "nothing to compress" instead of feeding the
 * whole file to the model.
 */
export function parseHistoryIntoTurns(historyContent: string): TurnEntry[] {
  const matches = [...historyContent.matchAll(TURN_PATTERN)]

  return matches.map((match, i) => {
    const start = (match.index ?? 0) + match[0].length
    const end = matches[i + 1]?.index ?? historyContent.length

    return {
      // The capture group is `\d+`, so this cannot be NaN.
      turnNumber: Number.parseInt(match[1] ?? '0', 10),
      location: match[2] ?? '',
      content: historyContent.slice(start, end).trim(),
    }
  })
}

/**
 * Chunk turns into fixed-size batches, in order.
 *
 * The last batch is short rather than padded or merged, so compressing 7 turns
 * at a batch size of 3 produces three sections, the last covering one turn.
 * A `batchSize` of 0 or less would loop forever in Python's `range(0, n, 0)`
 * equivalent; it is rejected here instead.
 */
export function groupTurnsIntoBatches(
  turns: TurnEntry[],
  batchSize: number = BATCH_SIZE,
): TurnEntry[][] {
  if (!Number.isInteger(batchSize) || batchSize < 1) {
    throw new RangeError(`batchSize must be a positive integer, got ${batchSize}`)
  }

  const batches: TurnEntry[][] = []
  for (let i = 0; i < turns.length; i += batchSize) {
    batches.push(turns.slice(i, i + batchSize))
  }
  return batches
}

/**
 * Render a batch back into the `## Turn N - Location` form it was parsed from.
 *
 * Round-tripping rather than inventing a new format is deliberate: the
 * summarizer's own config describes history in these terms, and its output
 * heading style has to match what `getHistoryBySubtitle` later parses.
 *
 * Each entry is followed by a blank line — including the last, so the string
 * ends with a newline. Python's `"\n".join` over a trailing `""` produces
 * exactly that.
 */
export function formatBatchForSummarizer(batch: TurnEntry[]): string {
  const lines: string[] = []
  for (const entry of batch) {
    lines.push(`## Turn ${entry.turnNumber} - ${entry.location}`)
    lines.push(entry.content)
    lines.push('')
  }
  return lines.join('\n')
}

/**
 * The instruction wrapped around one batch.
 *
 * Reproduced verbatim from `history_compression_service.py:169-180`, trailing
 * ellipsis and all: it is a prompt, and paraphrasing it changes the output of
 * every compression run.
 */
export function buildSummarizerPrompt(batchContent: string): string {
  return `Please compress the following turn entries into a single consolidated section.

## Turn Entries to Compress
${batchContent}

## Instructions
1. Create a meaningful subtitle in square brackets that captures the key event/theme
2. Write a concise summary that preserves important events, characters, and outcomes
3. Output ONLY the consolidated section in this format:

## [meaningful_subtitle_here]
Your consolidated summary here...`
}

// ============================================================================
// The real summarizer
// ============================================================================

/**
 * A {@link Summarize} backed by the Agent SDK.
 *
 * Two details are worth stating rather than discovering:
 *
 * - **No tools, no MCP servers.** `History_Summarizer` has no entry under
 *   `agents:` in `agents/group_gameplay/group_config.yaml`, so it inherits only
 *   the group's `disabled_tools` and ends up with nothing enabled. Building the
 *   server set for it would produce an empty one at more cost, and would couple
 *   this module to the handler layer for no behavioural difference.
 * - **The session is keyed on room 0.** Python passes `room_id=0` for task
 *   tracking (`history_compression_service.py:124`); here that becomes a pool
 *   key of `room_0_agent_{id}`, so consecutive batches reuse one warm session —
 *   which also means each batch is summarised with the previous batches in
 *   context. That is Python's behaviour, and it is arguably why the sections
 *   read as a continuous history rather than as three unrelated summaries.
 */
export function createSummarizer(
  pool: SessionPool,
  options: { useSonnet?: boolean } = {},
): Summarize {
  const runner = new TurnRunner(pool)
  const useSonnet = options.useSonnet ?? getSettings().useSonnet

  return async (request) => {
    let responseText = ''

    for await (const event of runner.run({
      roomId: request.roomId,
      agentId: request.agent.id,
      agentName: request.agent.name,
      content: request.userMessage,
      // Nothing the summarizer says is shown to the player; the section it
      // returns is written to a file, not streamed into the room.
      hidden: true,
      options: {
        systemPrompt: request.systemPrompt,
        mcpServers: {},
        toolNames: [],
        useSonnet,
      },
    })) {
      if (event.type === 'content_delta') {
        responseText += event.delta
      } else if (event.type === 'stream_end' && event.responseText) {
        // The accumulated deltas and the final text are the same string in the
        // normal case; the assignment matters when the turn produced no deltas
        // at all, which is what a non-streaming result message looks like.
        responseText = event.responseText
      }
    }

    return responseText.trim() || null
  }
}

// ============================================================================
// Service
// ============================================================================

export class HistoryCompressionService {
  private readonly worlds: WorldService
  private readonly cache: MtimeCache
  private readonly configs: AgentConfigService

  /**
   * @param summarize The model call. Required, and not defaulted to
   *   {@link createSummarizer}, because that would need a `SessionPool` this
   *   class has no business owning — and would let a test silently spawn a CLI.
   * @param worldsDir Root of the `worlds/` tree.
   * @param configs Reader for the `agents/` tree, for the summarizer's own
   *   character files.
   */
  constructor(
    private readonly summarize: Summarize,
    worldsDir: string = getSettings().paths.worldsDir,
    configs: AgentConfigService = new AgentConfigService(),
  ) {
    // The cache is held rather than left inside WorldService because
    // `compressHistory` rewrites `history.md` through `fs` directly and has to
    // drop that path's entry — two writes inside one millisecond would
    // otherwise leave the emptied file invisible to the next `loadHistory`.
    this.cache = new MtimeCache()
    this.worlds = new WorldService(worldsDir, this.cache)
    this.configs = configs
  }

  /**
   * Compress a world's `history.md` and clear it.
   *
   * Three outcomes, all reported as `success: true` except the last:
   *
   * - nothing to compress (missing, or still just the `# World History` header);
   * - no turn entries found (a file with content but no headings);
   * - every batch failed, which is the only `success: false` — a *partial*
   *   failure is reported as success with a lower `sections_created`, because
   *   the sections that did land have been written and `history.md` is about to
   *   be cleared regardless.
   *
   * **That last point is the sharp edge.** Clearing `history.md` after a partial
   * run permanently discards the turns whose batches failed. Python does the
   * same (`history_compression_service.py:293-320`); it is called out here
   * because it looks like a bug and changing it would change which text a
   * player's world retains.
   */
  async compressHistory(
    db: Db,
    worldName: string,
    batchSize: number = BATCH_SIZE,
  ): Promise<CompressionResult> {
    const historyContent = this.worlds.loadHistory(worldName)

    if (!historyContent || historyContent.trim() === EMPTY_HISTORY_MARKER) {
      return emptyResult('No history to compress')
    }

    const turns = parseHistoryIntoTurns(historyContent)
    if (turns.length === 0) {
      return emptyResult('No turn entries found in history')
    }

    const batches = groupTurnsIntoBatches(turns, batchSize)
    logger.info(
      `Compressing ${turns.length} turns into ${batches.length} batches for world '${worldName}'`,
    )

    const summarizer = this.findSummarizerAgent(db)
    const compressedSections: string[] = []

    for (const batch of batches) {
      // Sequential, not `Promise.all`: the batches share one warm session (see
      // `createSummarizer`) and, more importantly, the sections are appended in
      // order, so a concurrent run would have to reassemble them anyway.
      const section = summarizer ? await this.generateSection(summarizer, batch) : null
      if (section) {
        compressedSections.push(section)
      } else {
        logger.warning(
          `Failed to compress batch starting at turn ${String(batch[0]?.turnNumber)}`,
        )
      }
    }

    if (compressedSections.length === 0) {
      return {
        success: false,
        turns_compressed: 0,
        sections_created: 0,
        message: 'Failed to generate any compressed sections',
      }
    }

    this.appendConsolidatedSections(worldName, compressedSections)
    logger.info(
      `Written ${compressedSections.length} sections to consolidated_history.md`,
    )

    this.resetHistory(worldName)
    logger.info(`Cleared history.md for world '${worldName}'`)

    return {
      success: true,
      turns_compressed: turns.length,
      sections_created: compressedSections.length,
      message: `Compressed ${turns.length} turns into ${compressedSections.length} sections`,
    }
  }

  /**
   * The `History_Summarizer` row, or `null` with a warning.
   *
   * Scoped to the `gameplay` group, which is what makes this immune to a world
   * character someone happened to name `History_Summarizer`. Python filters the
   * group in SQL and the name in Python (`history_compression_service.py:106-117`);
   * one `WHERE` is the same query.
   */
  private findSummarizerAgent(db: Db): Agent | null {
    const agent =
      db
        .select()
        .from(agents)
        .where(and(eq(agents.group, SUMMARIZER_GROUP), eq(agents.name, SUMMARIZER_NAME)))
        .get() ?? null

    if (!agent) logger.warning(`${SUMMARIZER_NAME} agent not found in database`)
    return agent
  }

  /**
   * Summarise one batch. `null` on any failure, which costs that batch and not
   * the run.
   *
   * The system prompt is rebuilt from the agent's config folder rather than read
   * from the `system_prompt` column, so an edit to the summarizer's character
   * files takes effect on the next compression without a reload — the
   * filesystem-primary rule, same as every other agent turn.
   */
  private async generateSection(summarizer: Agent, batch: TurnEntry[]): Promise<string | null> {
    try {
      const configData = this.configs.loadAgentConfig(summarizer.configFile)
      const systemPrompt = buildSystemPrompt(
        summarizer.name,
        configData ?? {
          inANutshell: null,
          characteristics: null,
          recentEvents: null,
          longTermMemorySubtitles: null,
        },
      )

      const raw = await this.summarize({
        agent: summarizer,
        systemPrompt,
        userMessage: buildSummarizerPrompt(formatBatchForSummarizer(batch)),
        roomId: 0,
      })

      // Trimmed here rather than left to the summarizer, so a blank section is
      // a batch failure for every implementation of {@link Summarize} and not
      // just for the SDK-backed one. Python trims in the same place
      // (`history_compression_service.py:216`).
      const section = raw?.trim() || null
      if (section) logger.info(`${SUMMARIZER_NAME} generated section: ${section.slice(0, 100)}...`)
      return section
    } catch (error) {
      logger.exception('Error generating compressed section', error)
      return null
    }
  }

  /**
   * Append sections to `consolidated_history.md`, creating it if needed.
   *
   * Existing content is right-trimmed before the join so repeated runs produce
   * exactly one blank line between sections rather than a growing gap.
   */
  private appendConsolidatedSections(worldName: string, sections: string[]): void {
    const file = join(this.worlds.getWorldPath(worldName), 'consolidated_history.md')

    const existing = existsSync(file) ? readFileSync(file, 'utf-8') : ''
    const added = sections.join('\n\n')
    const final = existing ? `${existing.trimEnd()}\n\n${added}` : added

    writeFileSync(file, final, 'utf-8')
  }

  /** Truncate `history.md` back to its header and drop the cached read. */
  private resetHistory(worldName: string): void {
    const file = join(this.worlds.getWorldPath(worldName), 'history.md')
    writeFileSync(file, `${EMPTY_HISTORY_MARKER}\n\n`, 'utf-8')
    this.cache.invalidate(file)
  }
}

function emptyResult(message: string): CompressionResult {
  return { success: true, turns_compressed: 0, sections_created: 0, message }
}
